canvas/async_enroll_gatherer.py

```diff
@@ -0,0 +1,179 @@
+                self.enrollments_in_progress[response_course_id]['enrollments'] += enrollment_dicts
+                self.enrollments_in_progress[response_course_id]['enroll_page_info'] = enrollment_page_info
+                self.enrollments_in_progress[response_course_id]['num_pages'] += 1
+
+        return None
+
+    def make_requests(self, course_ids: Sequence[int]) -> None:
+        with FuturesSession(max_workers=self.num_workers) as session:
+            responses = []
+            for course_id in course_ids:
+                # Prep params
+                params = copy.deepcopy(self.default_params)
+                params['variables']['courseID'] = course_id
+                if course_id in self.enrollments_in_progress.keys():
+                    enrollment_in_progress_dict = self.enrollments_in_progress[course_id]
+                    enroll_page_info = enrollment_in_progress_dict['enroll_page_info']
+                    params['variables']['enrollmentPageCursor'] = enroll_page_info['endCursor']
+
+                logger.debug(params['variables'])
+                response = session.post(self.complete_url, json=params)
+                responses.append(response)
+
+            for completed_response in as_completed(responses):
+                self.parse_enrollment_response(completed_response)
+
+    def generate_output(self) -> Tuple[pd.DataFrame, ...]:
+        logger.debug('generate_output')
+        enrollment_records = []
+        user_records = []
+        section_records = []
+
+        for course_id in self.enrollments_in_progress.keys():
+            enrollment_dicts = self.enrollments_in_progress[course_id]['enrollments']
+            for enrollment_dict in enrollment_dicts:
+                enrollment_record, user_record, section_record = unnest_enrollment(enrollment_dict)
+                enrollment_records.append(enrollment_record)
+                user_records.append(user_record)
+                section_records.append(section_record)
+
+        enrollment_df = pd.DataFrame(enrollment_records)
+        user_df = pd.DataFrame(user_records).drop_duplicates(subset=['canvas_id'])
+        section_df = pd.DataFrame(section_records).drop_duplicates(subset=['canvas_id'])
+        return (enrollment_df, user_df, section_df)
+
+    def gather(self) -> None:
+        logger.info('** AsyncEnrollGatherer')
+        logger.info('Gathering enrollment data asynchronously with GraphQL')
+
+        prev_course_id_lists = []
+        more_to_gather = True
+
+        while more_to_gather:
+            course_ids_to_process = sorted(self.get_course_ids_for_incomplete_enrollments())
+
+            if len(course_ids_to_process) == 0:
+                more_to_gather = False
+                logger.info('Enrollment records have been gathered')
+            else:
+                if (len(prev_course_id_lists) > 2) and (prev_course_id_lists[0] == course_ids_to_process) and (prev_course_id_lists[1] == course_ids_to_process):
```

Comment:
This check needs to be improved. I'm to find a clean way to enable re-tries. This MIGHT work some of the time, but is unreliable.

Reply:
I think I fixed this.